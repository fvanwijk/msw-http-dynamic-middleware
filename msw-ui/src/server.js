const { rest } = require('msw');
const { createServer } = require('@mswjs/http-middleware');
const { createHandlers } = require('msw-dynamic-http-middleware');
const pino = require('pino');

const logger = pino({
  prettyPrint: {
    translateTime: true,
    ignore: 'pid,hostname',
  },
});

const userSuccess = rest.get('/user', (req, res, ctx) => res(ctx.json({ name: 'frank' })));
const userError = rest.get('/user', (req, res, ctx) => res(ctx.status(500)));
const usersSuccess = rest.get('/users', (req, res, ctx) => res(ctx.json([{ name: 'frank' }])));
const usersError = rest.get('/users', (req, res, ctx) => res(ctx.status(500)));
const scenarios = {
  // Scenarios for one endpoint
  'user success': userSuccess,
  'user error': userError,
  'users success': usersSuccess,
  'users error': usersError,

  // Scenarios for multiple endpoints
  success: [userSuccess, usersSuccess],
  error: [userError, usersError],
};

const port = 9800;
const server = createServer(...createHandlers(scenarios, 'success'));
server.listen(port, () => logger.info(`MSW server running at http://localhost:${port}`));
