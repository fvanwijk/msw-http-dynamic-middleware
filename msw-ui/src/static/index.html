<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="UTF-8" />
    <link rel="icon" type="image/svg+xml" href="/assets/favicon.17e50649.svg" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>Vite App</title>
    <script type="module" crossorigin src="/assets/index.6ff39571.js"></script>
    <link rel="stylesheet" href="/assets/index.34039bd9.css">
  </head>
  <body>
    <div id="app">
      <div class="text-center">
        <h1>MSW UI</h1>
      </div>

      <h2>Set global scenario</h2>
      <div id="global-scenarios"></div>

      <h2>Set scenario per handler</h2>

      <table>
        <thead>
          <tr>
            <th>Method</th>
            <th>Path</th>
            <th>Scenario</th>
          </tr>
        </thead>
        <tbody id="scenarios-per-handler"></tbody>
      </table>
    </div>
    
  </body>
</html>
