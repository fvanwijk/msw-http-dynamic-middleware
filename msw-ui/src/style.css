@tailwind base;

@layer base {
  #app {
    @apply p-4;
  }

  h1 {
    @apply text-4xl font-bold;
  }
  h2 {
    @apply text-xl font-medium mb-4 mt-8;
  }

  button {
    @apply px-3 py-1 mr-2 bg-gray-300 hover:bg-gray-400 text-gray-800 rounded;
  }
  button.active {
    @apply bg-blue-700 hover:bg-blue-600 text-white;
  }

  table tr th {
    @apply text-gray-600 font-medium;
  }
  table tr td,
  table tr th {
    @apply px-4 py-2 text-left;
  }
  table tr td:first-child,
  table tr th:first-child {
    @apply pl-0;
  }
}

@tailwind components;
@tailwind utilities;
