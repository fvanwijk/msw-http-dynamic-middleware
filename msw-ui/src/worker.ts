import { rest, setupWorker } from 'msw';

const worker = setupWorker(
  rest.get('/scenario', (_, res, ctx) => {
    return res(
      ctx.json({
        'user success': rest.get('/user', (_, res, ctx) => res(ctx.json({ name: 'frank' }))),
        'user error': rest.get('/user', (_, res, ctx) => res(ctx.status(500))),
        'users success': rest.get('/users', (_, res, ctx) => res(ctx.json([{ name: 'frank' }]))),
        'users error': rest.get('/users', (_, res, ctx) => res(ctx.status(500))),
        success: [
          rest.get('/user', (_, res, ctx) => res(ctx.json({ name: 'frank' }))),
          rest.get('/users', (_, res, ctx) => res(ctx.json([{ name: 'frank' }]))),
        ],
        error: [
          rest.get('/user', (_, res, ctx) => res(ctx.status(500))),
          rest.get('/users', (_, res, ctx) => res(ctx.status(500))),
        ],
      }),
    );
  }),
  rest.put('/scenario', (_, res, ctx) => res(ctx.status(205))),
);

export default worker;
