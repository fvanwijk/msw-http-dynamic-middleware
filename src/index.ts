import pino from 'pino';
import { ResponseResolver, rest, RestHandler } from 'msw';
import { Path } from 'node-match-path';

export type Scenarios = Record<string, RestHandler | RestHandler[]>;

const logger = pino({ prettyPrint: {
  translateTime: true,
  ignore: 'pid,hostname'
} });

const defaultResolver: ResponseResolver = (_, res, ctx) => res(ctx.status(200));

const assertPath = (path: Path): path is string => {
  if (path instanceof RegExp) {
    throw new Error(`Only strings as path are supported`);
  }
  return true;
}

/**
 * Create REST endpoints (handlers) based on the given scenarios.
 * When a scenario is set using PUT /scenario, the path and method of the scenario handler are used as keys in the activeResolvers map and the resolver is used as value
 *
 * @param {*} scenarios an object of RestHandlers with scenario name as key.
 * @returns RestHandler[]
 */
export const createHandlers = (scenarios: Scenarios)  => {
  /* Store currently active resolvers by method and path, for example:
   * {
   *   '/user/': {
   *     GET: [resolver]
   *   }
   * }
   */
  const activeResolvers: Record<string, Record<string, ResponseResolver>> = {};

  return [
    // Create mock endpoints for all defined scenarios. Possible duplicates
    ...Object.values(scenarios).flatMap(handler => {
      const handlers = Array.isArray(handler) ? handler : [handler];
      
      return handlers.map(handler => {
        const { method, path, header } = handler.info;
        return rest[method.toLowerCase() as keyof typeof rest](path, (req, res, ctx) => {
          // Forward call to active resolver that comes from scenario or fall back to default resolver
          if (assertPath(path)) {
            let resolver = activeResolvers[path]?.[method];
            
            if (!resolver) {
              resolver = defaultResolver
              logger.info(`${header} (default resolver)`);
            } else {
              logger.info(header);
            }
            
            return resolver(req, res, ctx);
          }
        });
      })
    }),

    // Create endpoint to set mock for any endpoint
    rest.put('/scenario', (req, res, ctx) => {
      const scenarioName = (req.body as Record<string, any>)?.scenario;

      if (!scenarioName) {
        return res(
          ctx.status(400),
          ctx.text(`Please provide a scenario name in the request body. Example: { "scenario": "user success" }`),
        );
      }

      const handler = scenarios[scenarioName];
      if (!handler) {
        return res(ctx.status(400), ctx.text(`Scenario "${scenarioName}" does not exist`));
      }
      
      const handlers = Array.isArray(handler) ? handler : [handler];

      const headers = handlers.map(handler => {
        const { path, method, header } = handler.info;
        if (assertPath(path)) {
          if (!(path in activeResolvers)) {
            activeResolvers[path] = {};
          }
          
          // @ts-ignore resolver is protected but I don't care
          activeResolvers[path][method] = handler.resolver;
          return header;
        }
        return;
      })
      
      logger.info(`Set scenario "${scenarioName}" with resolvers for endpoints: ${headers.join(', ')}`)

      return res(ctx.status(205));
    }),
  ];
};
