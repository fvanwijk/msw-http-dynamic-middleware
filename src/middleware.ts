import path from 'path';
import express from 'express';

// Serve the ui dir, where the UI app is built to
export const middleware = express.static(path.join(__dirname, 'ui'));
