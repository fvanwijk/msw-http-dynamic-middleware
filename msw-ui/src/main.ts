import { RestHandler } from 'msw';
import './style.css';

/* Data */

const scenarios: Record<string, RestHandler | RestHandler[]> = await fetch('/api/scenario').then(res => res.json());

const scenariosPerHandler = Object.entries(scenarios)
  .filter(([_, handlers]) => !Array.isArray(handlers))
  // Group by endpoint (info.header)
  .reduce((acc, [scenarioName, { info }]) => {
    const key = info.header;
    const [method, path] = key.split(' ');
    if (!(key in acc)) {
      acc[key] = { method, path, scenarios: [] };
    }
    acc[key].scenarios.push(scenarioName);

    return acc;
  }, {} as Record<string, { method: string; path: string; scenarios: string[] }>);

const globalScenarios = Object.entries(scenarios).filter(([_, handlers]) => Array.isArray(handlers));

const createButton = (scenario: string) => {
  const button = document.createElement('button');
  button.innerText = scenario;
  button.addEventListener('click', () => {
    fetch('api/scenario', {
      method: 'PUT',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({ scenario }),
    });
  });
  return button;
};

/* Global scenarios */

const div = document.querySelector<HTMLDivElement>('#global-scenarios')!;
div.innerHTML = '';

globalScenarios.forEach(([scenario]) => {
  const button = createButton(scenario);
  div.appendChild(button);
});

/* Scenarios per handler table */

const createCell = (contents: string): HTMLTableDataCellElement => {
  var td = document.createElement('td');
  td.innerHTML = contents;
  return td;
};

const tbody = document.querySelector<HTMLTableSectionElement>('#scenarios-per-handler')!;
tbody.innerHTML = '';

Object.values(scenariosPerHandler).forEach(endpoint => {
  var tr = document.createElement('tr');

  const { method, path, scenarios } = endpoint;
  tr.appendChild(createCell(method));
  tr.appendChild(createCell(path));

  const td = document.createElement('td');
  scenarios.forEach(scenario => {
    td.appendChild(createButton(scenario));
  });
  tr.appendChild(td);

  tbody.appendChild(tr);
});
