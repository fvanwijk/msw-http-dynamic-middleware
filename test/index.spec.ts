import { rest } from 'msw';
import { createHandlers } from '../src';

const scenarios = {
  // Scenarios for one endpoint
  'user success': rest.get('/user', (req, res, ctx) => res(ctx.json({ name: 'frank' }))),
  'user error': rest.get('/user', (req, res, ctx) => res(ctx.status(500))),
  'users success': rest.get('/users', (req, res, ctx) => res(ctx.json([{ name: 'frank' }]))),
  'users error': rest.get('/users', (req, res, ctx) => res(ctx.status(500))),

  // Scenarios for multiple endpoints
  success: [
    rest.get('/user', (req, res, ctx) => res(ctx.json({ name: 'frank' }))),
    rest.get('/users', (req, res, ctx) => res(ctx.json([{ name: 'frank' }]))),
  ],
  error: [
    rest.get('/user', (req, res, ctx) => res(ctx.status(500))),
    rest.get('/users', (req, res, ctx) => res(ctx.status(500))),
  ],
};

describe('createHandlers', () => {
  it('should create handlers from scenarios', () => {
    const handlers = createHandlers(scenarios);
    const handlersInfo = handlers.map(({ info: { callFrame, ...info } }) => info);

    expect(handlersInfo).toHaveLength(10); // 8 scenarios + PUT /scenario + DELETE /scenario
    expect(handlersInfo).toMatchSnapshot();
  });
});
