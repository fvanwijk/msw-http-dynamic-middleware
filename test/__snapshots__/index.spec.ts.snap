// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createHandlers should create handlers from scenarios 1`] = `
Array [
  Object {
    "header": "GET /user",
    "method": "GET",
    "path": "/user",
  },
  Object {
    "header": "GET /user",
    "method": "GET",
    "path": "/user",
  },
  Object {
    "header": "GET /users",
    "method": "GET",
    "path": "/users",
  },
  Object {
    "header": "GET /users",
    "method": "GET",
    "path": "/users",
  },
  Object {
    "header": "GET /user",
    "method": "GET",
    "path": "/user",
  },
  Object {
    "header": "GET /users",
    "method": "GET",
    "path": "/users",
  },
  Object {
    "header": "GET /user",
    "method": "GET",
    "path": "/user",
  },
  Object {
    "header": "GET /users",
    "method": "GET",
    "path": "/users",
  },
  Object {
    "header": "PUT /scenario",
    "method": "PUT",
    "path": "/scenario",
  },
  Object {
    "header": "DELETE /scenario",
    "method": "DELETE",
    "path": "/scenario",
  },
]
`;
